### PR TITLE
[fix][ShardedDDP] Properly handle .eval() mode

### LINKED ISSUE #586: [ShardedDDP] Handle transition to eval + parameter change
## 🐛 Bug
Reported by @SeanNaren, after a successful training the switch to eval() is not properly taken into account, and the grads are marked as "waiting to be reduced" while they should not (we're in eval..)

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes eval mode not being properly handled by ShardedDDP, at best it kept all the grad hooks in place, at worst it could crash when there was a trainability change which should have been ignored. Add a unit test to catch that

Closes #586 
cc @SeanNaren  @ananthsub

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/nn/data_parallel/test_sharded_ddp_features.py [hunk 1]
```diff
@@ -262,6 +262,29 @@ def test_mixed_types():
...
+
+    model = _get_mlp()
+    model.train()
+    optimizer = OSS(params=model.parameters(), optim=torch.optim.SGD, lr=1e-3, momentum=0.99)
+    model = ShardedDataParallel(model, optimizer)
+    input_tensor = torch.rand((2, 2))
+    loss = model(input_tensor).sum()
+    loss.backward()  # make sure that the gradients are reduced
+
+    model = model.eval()
```
Comment (anj-s) [inline tests/nn/data_parallel/test_sharded_ddp_features.py:277]: Is there something else we should check? e.g that the weights haven't changed. This only checks that we don't throw an error.

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 2]
```diff
@@ -490,6 +486,9 @@ def _setup_backward_hooks(self) -> None:
         # Go through the parameters, attach the hook
         self._grad_accs = []
         self._manual_reduce = []
+        if not self.training:
```
Comment (anj-s) [inline fairscale/nn/data_parallel/sharded_ddp.py:489]: do we need to remove existing hooks in eval model? Just curious, otherwise we could move this to the top of the function.

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 3]
```diff
@@ -624,3 +623,19 @@ def _flush_reduce_calls(self) -> None:
                 bucket.sent = True
 
         self._consume_work_handles()
+
+    def _detect_train_change(self) -> bool:
```
Comment (anj-s) [inline fairscale/nn/data_parallel/sharded_ddp.py:627]: Nice!

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 4]
```diff
@@ -624,3 +623,19 @@ def _flush_reduce_calls(self) -> None:
...
+        # Optionally check whether the trainable parameters have changed
+        trainable_mask = list(map(_trainable, self._all_params))
+
+        # - one or more parameters trainability changed
+        trainability_changed = trainable_mask != self._reference_trainable_mask
+
+        # - the whole model is not trainable but we still have grad hooks
+        trainability_changed |= not self.training and len(self._grad_hooks) > 0
```
Comment (anj-s) [inline fairscale/nn/data_parallel/sharded_ddp.py:635]: does this mean that grad_hooks should be greater than 0 in eval model? Not sure I understand why this should be the case.

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 5]
```diff
@@ -624,3 +623,19 @@ def _flush_reduce_calls(self) -> None:
...
+
+        if trainability_changed:
+            logging.warning("ShardedDDP detected that the trainable params changed, updating the partitioning")
```
Comment (anj-s) [inline fairscale/nn/data_parallel/sharded_ddp.py:638]: update logging statement since "partitioning" is not clear in this function.

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 2 above)
Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:489]: I thought that it was better for correctness, in that if there's a .backward() left somewhere it still respects the eval() setting ? The documentation is not super clear, to me at least https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=eval#torch.nn.Module.train

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 4 above)
Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:635]: it was meant to detect that the trainability changed, ie. we're in eval() mode but there are grad_hooks in place so we should refresh ? it's tied to the question above, I'm not sure of the reference behavior here

Comment (anj-s) [reply fairscale/nn/data_parallel/sharded_ddp.py:635]: From my offline conversation with @blefaudeux to understand this better:
* We can't detect when a module switches from train->eval unless we use the presence of hooks as an indicator. 
* We refresh trainable 1) at the beginning 2) when params changes their requires_grad property 3) train<->eval switch. 

Thanks for the explanation @blefaudeux !

**Review by anj-s — APPROVED** [review]

**Comment by SeanNaren** [comment]
Fixes the issue upstream in Lightning, thanks so much for the quick fix @blefaudeux :)
